### DISCUSSION TITLE
Preview of version 4.0.0

### DISCUSSION THREAD

**Discussion by waydabber** [MonitorControl/MonitorControl#596] [Show and tell]
**Update: this discussion is now locked. We can continue the discussion in the relevant release discussion thread after the Pull Request https://github.com/MonitorControl/MonitorControl/pull/629 is merged. Thanks to all who gave feedback and encouragement!**

---

Hey Everybody,

I just want to give an update on the upcoming version. This is going to be a rather big release again with a lot of changes. Some of you might have an update fatigue so we won't be releasing this until all things are completed and settled. After this update the pace of changes will decrease, but right now we have a huge backlog of feature requests.

**A note to our beloved translators:** A lot of text will change or will be added in this release yet again. We'll let you know when it is time to translate. Please be ready. Thank you! :)

The following change log and screenshots are as of **30/09/2021**.

### Added/improved functionality

- Added proper support for controlling Apple displays.
- Added option to show/hide brightness slider.
- Added option to show brightness slider for internal display and apple displays (enabled by default).
- Replication of built-in and Apple display brightness to corresponding brightness slider.
- Added suffix to similarly named displays for better differentiation.
- Option to disable slider snapping for finer control + disable slider snapping by default.
- Added option to show slider tick marks for better accuracy.
- Added option to use window focus instead of mouse to determine which display to control.
- Ctrl+Command+Brightness now controls external displays only (Ctrl+Brightness continues to control internal display only)
- Added separate tab for menu options.
- Added option to restore last saved values upon startup.
- Added option for audio device name matching for display volume control selection.
- Separated option to change all screens for brightness and volume.
- Added option for keyboard fine scale for brightness.
- Added option for keyboard fine scale for volume.
- Added version check upon startup for mandatory preferences reset upon downgrade or incompatible previous version + notification about this.
- Added implementation for Command+F1 macOS shortcut to enable/disable mirroring.
- Added safer 'Assume last saved settings are valid' option as default instead of startup DDC read (or restore).
- Streamlined preference panes, 'Show advanced settings' now affect all tabs. This leads to a better and safer first timer experience (especially because of the influx of many new features).
- Added a Quit button to Preferences if menu is hidden (it was not passible to quit the application until this time in this mode only by re-enabling the menu).
- Lowered default first-run volume DDC default from 75% to 15% if read is not possible or disabled to prevent unexpectedly loud sound. 
- Added slider skew setting on a per control basis to have the ability to manipulate DDC slider balance and OSD scale if display control is not linear.
- Added the ability to set min. and max. DDC bounds on a per display, per control basis.
- Audio device name override option for a display (manually assign a specific audio device to a display).
- Advanced setting to invert DDC control range (some displays have the scale reversed).
- Advanced setting to remap DDC control code (some displays have contrast and brightness mixed up).
- Ability to mark a DDC control as available or unavailable in advanced settings under Displays.
- Ability to automatically hide menu icon if there is no slider present in the menu.
- Option to show slider percentage for more precision.
- Option to set combined or separate OSD scale when combined hardware+software brightness is used.
- Apple like smooth brightness change (both for software, hardware, mixed and DisplayServices).
- Added support for DisplayLink, AirPlay, Sidecar, screen sharing etc. using window shades (this is an inferior technique to the existing software implementation - gamma control - but still better than nothing). Disabled for any kind of mirroring setups. _(Only on Big Sur and above)_
- Brightness change synchronization from Built-In and Apple displays to other displays. This makes Touch Bar, Ambient light sensor, Control Center and System Preferences induced changes affect all displays. Synchronization uses a sophisticated indirect delta method + the user can intervene and adjust individual screen brightness at any time to easily compensate mismatching native brightness levels.
- Preferences pane tab selector has a simpler look on Catalina.
- All menu sliders are now scrollable using a magic mouse/trackpad swipes or mouse wheel.
- Added option for menu to show only items that are relevant to display which shows the menu currently.
- Added option to enable combined sliders (note: this option combined with enabled Apple/built-in display syncing and enabled 'change all' keyboard settings finally provides full synchronized control of all displays).
- Combined sliders can now display multiple displays when keyboard and brightness syncing is not enabled. _(Only on Big Sur and above)_
- Redesigned sliders to look like Big Sur/Monterey Control Center's sliders. _(Only on Big Sur and above)_
- Quit and Preferences... are now icons for a much cleaner look. _(Only on Big Sur and above)_
- Added option to change additional menu options style or hide them. _(Only on Big Sur and above)_
- Multiple displays are now in nice Big Sur styled blocks - no more ugly separators. _(Only on Big Sur and above)_
- Added customizable gamma/ddc switchover point for combined brightness in the advanced section of Displays.
- Added comma separated list for control code override to enable edge cases like controlling Brightness and Contrast at the same time (use VCP list entry `10, 12` for that)
- Contrast can now be controlled from keyboard via control+option+command+brightness up/down.
- Custom keyboard shortcuts for brightness, contrast, volume and mute
- Check for update feature for automatic updates.

### Other under the hood changes and bug fixes

- Standardized internal scale among various displays and DDC ranges for ranged controls.
- Uses the new internal scale for combined hardware-software brightness mode.
- Migrated scales to internal float representation to prevent loss of fine detail on transformations.
- Fixed double sound when muting multiple external displays at the same time.
- Fixed lack of initial volume configuration if slider is not shown in menu.
- Fixed wrong settings being applied to a display when replaced on Apple Silicon (UserDefaults preferences are now tied to specific display strings instead of CGDirectDisplayID - which is no longer semi-unique on arm64).
- A lot of refactoring, streamlining and general optimizations.

## Known Issues

- N/A

## Here are some screenshots of the current state of the app:

Note: the screenshots reflect `Advanced options` enabled. With this disabled the app looks much less intimidating.

<div align="center">

<img width="691" alt="Screen Shot 2021-09-29 at 13 00 00" src="https://user-images.githubusercontent.com/37590873/135257000-2bd6d79a-5a66-47a8-9381-d1a0c9a3d3f7.png">
<img width="803" alt="Screen Shot 2021-09-30 at 14 24 28" src="https://user-images.githubusercontent.com/37590873/135454345-ef331fa9-a351-4a8c-a942-d1fd94025131.png">
<img width="803" alt="Screen Shot 2021-09-28 at 21 35 19" src="https://user-images.githubusercontent.com/37590873/135154129-644a6ba4-3e68-4abd-ad0f-2cb20d2c0cd9.png">
<img width="803" alt="Screen Shot 2021-09-28 at 21 34 40" src="https://user-images.githubusercontent.com/37590873/135154016-1df39c6d-deb9-4656-b475-821108bf2b81.png">
<img width="803" alt="Screen Shot 2021-09-30 at 14 24 24" src="https://user-images.githubusercontent.com/37590873/135454404-514a94eb-e4dc-4bdd-a101-cf143570b4b7.png">
<img width="803" alt="Screen Shot 2021-09-27 at 17 30 55" src="https://user-images.githubusercontent.com/37590873/134939705-77b44218-3e49-4f1f-890d-eb65b492c65a.png">
<img width="803" alt="Screen Shot 2021-09-27 at 11 41 33" src="https://user-images.githubusercontent.com/37590873/134885061-c0bdf90f-971c-47ce-bd85-bc870ed05a77.png">

</div align="center">

**Comment by Muirium** [comment]
Looking really good!

I'd love to get [keyboard controls for Contrast](https://github.com/MonitorControl/MonitorControl/discussions/590) as well as Brightness. My Dell P2415Q is old-school, that way: both need tweaked a little differently. I'm thinking F1 & F2 for Brightness and Shift+F1 & Shift+F2 for Contrast control, on [my vintage mechanical keyboards](https://deskthority.net/download/file.php?id=27872). Configurable keys are just what the display doctor ordered!

**Reply by waydabber** [reply]
Hi @Muirium - can you describe in more depth why the P2415Q needs both brightness and contrast control? I am asking this because in my mind contrast control is not very useful on modern flat panel displays with digital connection - it is better to calibrate the display contrast so the full color color range is displayed without capping (there is a proper calibration method for that) and then change only the backlight brightness + use software gammatable dimming if you need to go darker than the minimum possible setting (this is superior to the contrast control most display provides). But your situation with the P2415Q might be special, I am not familiar with that display model. :)

**Reply by Muirium** [reply]
Hey @waydabber!

It's a 4K Dell display from 2014. Nice high DPI—just 24 inches—but definitely retro controls!

Adjusting Brightness alone leads to a washed out mess at either end of its range. Adjusting Contrast seems to affect the LED backlight level too: it goes WAY BRIGHTER when both both Brightness AND Contrast are maxed out! The sweet spot varies with how much room brightness you're dealing with, and what image / video you're looking at. Sometimes Contrast needs to be as much as 50% higher than Brightness, other times they need to be more equal.

Before version 3.0 I was using [Lunar](https://lunar.fyi) to control it, but the GUI was driving me nuts! This display really doesn't like being second guessed with "smart" curves. Lunar's manual mode was okay, but very fiddly to dial in the small adjustments I need.

I think Dell was trying to mimic the feeling of twiddling Brightness and Contrast dials on CRTs with these displays. I had an even older 20 inch 1680x1050 LCD of theirs before this which felt exactly the same. Nothing at all like Apple's panels.

**Reply by waydabber** [reply]
Hi, thanks for the explanation! I recommend you adding a Feature Request for it. I think it is possible to add this functionality + an advanced setting to enable this.

**Reply by Muirium** [reply]
Done:

#603 

It feels like a natural fit for MonitorControl. I remember looking for this feature when first setting it up on my M1.

**Comment by Muirium** [comment]
Good idea incrementing to version 4. The amount of new features on display here is not only major, it's magnificent!

**Comment by derpyjoystick** [comment]
First, thank you so much for developing this. May seem like a small utility but having to change the display's brightness using the onboard controls that require 4 inputs just to get too was insufferable!

Specifically with v4's new combination of software and hardware dimming, would it possible to set the software dimming later? Currently, it hits software dimming before the HUD reaches 50% using the keyboard shortcuts making the process reach an overall darker state sooner. It would be nice if it occurred later and the process focused more on keeping the display brighter, where software dimming could kick in near the last 3-4 increments of the HUD.
![iu-1](https://user-images.githubusercontent.com/2991193/134823664-886e8f58-2eee-4f55-a6f2-bbb3ea686038.jpeg)

The yellow arrow in the attachment is where the display reaches 0% hardware brightness. The fuchsia arrow is the suggested point where the display would reach 0, and the rest would be software brightness kicking in.

In any case, it's a truly lovely feature as 4K displays never really reach low brightness let alone a state of what appears as "off." Thank you for including it!

**Reply by waydabber** [reply]
Hi @ultratiem, yes, great idea, was planning to add a slider to define the point of switch in advanced display prefs, the backend is ready for that.

Note: you can disable this feature or you can enable OSD scale/Separate scales... which gives you 100% OSD scale for backlight brightness control and kicks gammatable dimming after that (this option is available if Advanced preferences is enabled).

**Reply by derpyjoystick** [reply]
Thanks for the workarounds. I'll use them in the interim but do actually like having the entire process work elegantly together. It's a truly beautiful feature that offers a very delightful Ux. Absolute 🤌

And thanks again for the little app. It's a gem 🙏

**Reply by waydabber** [reply]
Hi, I added a slider `Combined dimming switchover point` for it. You can try if it works as intended. Also if you can test all other aspects of the app and report bugs, I'll be very thankful! :)

**Reply by derpyjoystick** [reply]
Will do thanks for fleshing out the feature! So far, everything I've tested is bug free but I'll keep an eye out. Just an FYI but the DB linked to the unsigned build isn't working.

**Reply by waydabber** [reply]
Hmm. You mean the unsigned download link is not working? It seems to work for me, but maybe I miss something. But it's safer to build it yourself anyway than downloading and running unsigned stuff from the internet. :)

**Comment by acousticbiker** [comment]
Nice! I had my fingers crossed to see auto-updates - is that on the roadmap?

**Reply by waydabber** [reply]
Hi @acousticbiker, yes it is. You can see the current yet unimplemented feature requests here:

https://github.com/MonitorControl/MonitorControl/issues?q=is%3Aopen+is%3Aissue+-label%3A%22Status%3A+Done%22

I am not sure what will make it into 4.0.0 but these should be implemented sooner or later.

**Comment by Muirium** [comment]
Playing with the version 4 preview on my Dell P2415Q just now. Nice work! The new features are sweet.

Magic Mouse side scrolling over the Brightness and Contrast sliders does work. Hooray! Quite nicely done. Slider Snapping throws it off quite a bit, though. With slider snapping enabled: the sliders are not nearly as smooth as Apple's native sliders. But now I look around for examples of Apple's own sliders with detents (snapping), I see they don't support mouse scrolling anyway! (Two examples: Settings > Sound > Balance slider and Settings > Accessibility > Spoken Content > Speaking Rate slider.) I know how I want them to feel, but Apple's don't even side scroll at all!

So I'm pretty chuffed with how v4 is coming. (Contrast keys and I'll be delighted!)

**Comment by Muirium** [comment]
A quick bug report on the v4.0.0 build 6085, on M1 MacBook Air. No external display connected:

![Screenshot 2021-09-28 at 12 58 55 pm](https://user-images.githubusercontent.com/221646/135082842-d972c90c-9b76-4520-a567-27e5beaba591.jpg)

And that's with Menu icon set to: "Only if at least one slider is present" and Brightness control: "Enable for Apple branded and built-in displays as well" deselected. Obviously the intention is to hide the menu completely when the Air's built in display is the only one connected.

**Reply by waydabber** [reply]
Yes, you are right. I did replicate the issue, I'll fix it. Nice find, thank you!

**Reply by waydabber** [reply]
Also menu options are not visible when there are no sliders present and the menu options are in icon mode. I reimplemented how menus are working and still there are some issues apparently.

**Reply by waydabber** [reply]
I fixed this now and uploaded a build. Please let me know if you find anything else. Thank you!

**Reply by Muirium** [reply]
Trying it now: pulling in and out my Thunderbolt Dock (with display attached via DisplayPort) works just great. MonitorControl's menu now hides and reappears just as it should. Nice one!

Next up: toggling "Combine hardware and software dimming" throws my Dell P2415Q to 100% hardware brightness, both ways! The display is instantly MUCH BRIGHTER whenever I touch that checkbox. Perhaps try to keep the screen brightness equivalent when flipping that switch? Or at least closer than the giant leap it can make now.

Besides that jarring state change, the feature works really quite nicely on my setup. The Brightness slider feels natural in both (software dimming) enabled and disabled states! Pretty cool.

I really like the precision of side scrolling single %s at a time on both Brightness and Contrast sliders. Especially now it's nighttime here. Version 4 is superb!

**Reply by waydabber** [reply]
Glad you like it so far! :)

**Comment by waydabber** [comment]
I'll lock this thread as 4.0.0 is about to be available as an official pre-release version. We'll gather feedback in the release discussion.

Pull Request: https://github.com/MonitorControl/MonitorControl/pull/629

**Comment by JoniVR** [comment]
v4.0.0-beta1 is out!
